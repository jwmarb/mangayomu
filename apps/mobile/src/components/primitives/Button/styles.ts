import { ViewStyle } from 'react-native';
import type {
  ButtonVariants,
  ButtonColors,
} from '@/components/primitives/types';
import { createStyles } from '@/utils/theme';

export const variants = (color: ButtonColors, state: 'enabled' | 'disabled') =>
  createStyles<Record<ButtonVariants, ViewStyle>>((theme) => ({
    contained: {
      backgroundColor:
        state === 'disabled'
          ? theme.palette.action.disabled
          : theme.palette[color].main,
    },
    outlined: {
      borderColor:
        state === 'disabled'
          ? theme.palette.action.disabled
          : theme.palette[color].main,
    },
    text: {},
  }));

export const styles = (variant: ButtonVariants, color: ButtonColors) =>
  createStyles((theme) => ({
    button: {
      paddingHorizontal: theme.style.size.xxl,
      paddingVertical: theme.style.size.l,
    },
    container: {
      borderWidth: 1.5,
      borderRadius: theme.style.borderRadius,
      borderColor:
        variant === 'outlined' ? theme.palette[color].main : 'transparent',
      overflow: 'hidden',
    },
    pressable: {
      gap: theme.style.size.m,
      alignItems: 'center',
      flexDirection: 'row',
    },
    text: {
      flexGrow: 1,
    },
  }));
