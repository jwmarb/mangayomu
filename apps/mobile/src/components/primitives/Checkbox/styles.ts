import { createStyles } from '@/utils/theme';

export const styles = createStyles((theme) => ({
  pressableCheckbox: {
    flex: 0,
    width: 36,
    height: 36,
    alignItems: 'center',
    justifyContent: 'center',
  },
  checkboxChecked: {
    padding: theme.style.size.s * 0.5,
    borderColor: 'transparent',
    borderRadius: theme.style.borderRadius.m,
    borderWidth: theme.style.borderWidth.s,
    backgroundColor: theme.palette.primary.main,
    flex: 0,
  },
  checkbox: {
    padding: theme.style.size.s * 0.5,

    borderColor: theme.palette.divider,
    borderRadius: theme.style.borderRadius.m,
    borderWidth: theme.style.borderWidth.s,
    flex: 0,
  },
  uncheckedIcon: {
    opacity: 0,
  },
}));
