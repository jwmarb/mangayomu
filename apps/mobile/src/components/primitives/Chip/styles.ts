import { ChipColors } from '@/components/primitives/types';
import { createStyles } from '@/utils/theme';

export const styles = (color: ChipColors | 'null') =>
  createStyles((theme) => ({
    chip_outlined: {
      borderColor:
        color === 'null' ? theme.palette.divider : theme.palette[color].main,
      borderRadius: theme.style.borderRadius / 3,
      borderWidth: 1.5,
      flexDirection: 'row',
      overflow: 'hidden',
    },
    chip_filled: {
      backgroundColor:
        color === 'null'
          ? theme.palette.divider
          : theme.palette[color][theme.mode === 'dark' ? 'dark' : 'light'],
      borderRadius: theme.style.borderRadius / 3,
      borderWidth: 1.5,
      flexDirection: 'row',
      overflow: 'hidden',
    },
    chipPressable: {
      flex: 0,
      flexDirection: 'row',
      paddingVertical: theme.style.size.m,
      paddingHorizontal: theme.style.size.l,
      alignItems: 'center',
      gap: theme.style.size.s,
    },
  }));
