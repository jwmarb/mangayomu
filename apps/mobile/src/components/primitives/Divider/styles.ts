import { createStyles } from '@/utils/theme';

export const styles = createStyles((theme) => ({
  horizontal: {
    flexGrow: 1,
    height: 1.5,
    backgroundColor: theme.palette.divider,
  },
  vertical: {
    flexGrow: 1,
    width: 1.5,
    backgroundColor: theme.palette.divider,
  },
}));
