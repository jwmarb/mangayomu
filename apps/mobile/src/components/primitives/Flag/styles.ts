import { FLAG_HEIGHT, FLAG_WIDTH } from '@/components/primitives/Flag';
import { createStyles } from '@/utils/theme';

export const styles = createStyles(() => ({
  flag: {
    width: FLAG_WIDTH,
    height: FLAG_HEIGHT,
  },
}));
