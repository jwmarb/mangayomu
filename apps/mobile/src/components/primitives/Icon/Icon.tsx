import MaterialCommunityIconNames from 'react-native-vector-icons/glyphmaps/MaterialCommunityIcons.json';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';
import {
  Image,
  ImageStyle,
  StyleProp,
  TextProps,
  TextStyle,
} from 'react-native';
import React from 'react';
import Animated, { AnimatedProps } from 'react-native-reanimated';
import { sizes } from '@/components/primitives/Icon/styles';
import {
  IconSizes,
  TextColorTypes,
  TextColors,
} from '@/components/primitives/types';
import useStyles from '@/hooks/useStyles';
import useContrast from '@/hooks/useContrast';
import { composedColors } from '@/components/primitives/Text';

export type IconProps = BaseProps & (FontIconProps | ImageIconProps);

type BaseProps = Pick<TextProps, 'style'> & {
  color?: TextColors;
  colorType?: TextColorTypes;
  size?: 'large' | 'medium' | 'small' | number;
  contrast?: boolean;
  testID?: string;
};
type FontIconProps = {
  type: 'icon';
  name: keyof typeof MaterialCommunityIconNames;
};

type ImageIconProps = {
  type: 'image';
  uri: string;
};

export const composedSizes = (['icon', 'image'] as const).reduce(
  (prev, curr) => {
    prev[curr] = sizes(curr);
    return prev;
  },
  {} as Record<Pick<IconProps, 'type'>['type'], Record<IconSizes, TextStyle>>,
);

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function Icon(props: IconProps, ref: React.ForwardedRef<any>) {
  const {
    color: colorProp = 'textPrimary',
    colorType = 'main',
    size: sizeProp = 'medium',
    contrast: contrastProp,
    style: styleProp,
    ...rest
  } = props;
  const contrast = useContrast(contrastProp);
  const size =
    typeof sizeProp === 'string'
      ? composedSizes[rest.type][sizeProp]
      : undefined;
  const color = useStyles(composedColors[colorType], contrast)[colorProp];
  const style =
    rest.type === 'icon' ? [size, color, styleProp] : [size, styleProp];
  switch (rest.type) {
    case 'icon':
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      return (
        <MaterialCommunityIcons
          {...rest}
          ref={ref}
          name={rest.name}
          size={typeof sizeProp === 'number' ? sizeProp : undefined}
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          style={style as any}
        />
      );
    case 'image':
      return (
        <Image
          {...rest}
          ref={ref}
          src={rest.uri}
          style={style as StyleProp<ImageStyle>[]}
        />
      );
  }
}

const ForwardedIcon = React.forwardRef(Icon);

export const AnimatedIcon = Animated.createAnimatedComponent(
  ForwardedIcon,
) as React.FC<IconProps & AnimatedProps<IconProps>>;

export default ForwardedIcon as typeof Icon;
