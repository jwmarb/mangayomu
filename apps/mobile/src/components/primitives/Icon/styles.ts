import { StyleSheet, TextStyle } from 'react-native';
import { IconSizes } from '@/components/primitives/types';
import { IconProps } from '@/components/primitives/Icon/Icon';

export const sizes = (type: Pick<IconProps, 'type'>['type']) =>
  StyleSheet.create<Record<IconSizes, TextStyle>>({
    large:
      type === 'icon'
        ? {
            fontSize: 32,
          }
        : {
            width: 32,
            height: 32,
          },
    medium:
      type === 'icon'
        ? {
            fontSize: 24,
          }
        : {
            width: 24,
            height: 24,
          },
    small:
      type === 'icon'
        ? {
            fontSize: 16,
          }
        : {
            width: 16,
            height: 16,
          },
  });
