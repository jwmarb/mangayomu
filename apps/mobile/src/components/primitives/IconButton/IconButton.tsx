import { View } from 'react-native';
import React from 'react';
import { iconButton } from '@/components/primitives/IconButton/styles';
import { ButtonProps } from '@/components/primitives/Button/Button';
import Pressable from '@/components/primitives/Pressable';
import Icon, { IconProps } from '@/components/primitives/Icon';
import { createThemedProps } from '@/utils/theme';
import {
  ICON_BUTTON_COLORS,
  IconButtonColors,
} from '@/components/primitives/types';
import useThemedProps from '@/hooks/useThemedProps';
import useContrast from '@/hooks/useContrast';

export type IconButtonProps = Omit<ButtonProps, 'title' | 'color'> & {
  icon: React.ReactElement<IconProps, typeof Icon>;
  color?: IconButtonColors;
};

const clonedIconProps = (color: IconButtonColors) => ({
  style: { textAlign: 'center' as const },
  color: color === 'default' ? undefined : color,
});

const composedClonedIconProps = ICON_BUTTON_COLORS.reduce((prev, curr) => {
  prev[curr] = clonedIconProps(curr);
  return prev;
}, {} as Record<IconButtonColors, Partial<IconProps>>);
const pressableProps = (color: IconButtonColors) =>
  createThemedProps((theme) => ({
    android_ripple: {
      borderless: true,
      color: color === 'default' ? undefined : theme.palette[color].ripple,
    },
  }));
const composedPressableProps = ICON_BUTTON_COLORS.reduce((prev, curr) => {
  prev[curr] = pressableProps(curr);
  return prev;
}, {} as Record<IconButtonColors, ReturnType<typeof pressableProps>>);

export default function IconButton(props: IconButtonProps) {
  const { icon, contrast: contrastProp, color = 'default', ...rest } = props;
  const contrast = useContrast(contrastProp);
  const { android_ripple } = useThemedProps(
    composedPressableProps[color],
    contrast,
  );
  return (
    <View style={iconButton.container}>
      <Pressable
        style={iconButton.pressable}
        android_ripple={android_ripple}
        {...rest}
      >
        {React.cloneElement(icon, composedClonedIconProps[color])}
      </Pressable>
    </View>
  );
}
