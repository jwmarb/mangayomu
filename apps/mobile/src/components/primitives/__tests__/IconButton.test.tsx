import { render, userEvent } from '@testing-library/react-native';
import Icon from '@/components/primitives/Icon';
import IconButton from '@/components/primitives/IconButton';
import { iconButton } from '@/components/primitives/IconButton/styles';

beforeAll(() => {
  jest.useFakeTimers();
});

afterAll(() => {
  jest.useRealTimers();
});

test('renders correctly', () => {
  const tree = render(
    <IconButton icon={<Icon type="icon" name="ab-testing" />} />,
  );
  expect(tree).toMatchSnapshot();
});

test('proper interaction', async () => {
  const callback = jest.fn();
  const user = userEvent.setup();
  const { getByTestId } = render(
    <IconButton
      onPress={callback}
      icon={<Icon type="icon" name="ab-testing" />}
      testID="iconbtn"
    />,
  );

  await user.press(getByTestId('iconbtn'));

  jest.runOnlyPendingTimers();

  expect(callback).toHaveBeenCalled();
});

test('has correct styling', () => {
  const { root } = render(
    <IconButton size="small" icon={<Icon type="icon" name="ab-testing" />} />,
  );
  expect(root).toHaveStyle(iconButton.containerSmall);
});
