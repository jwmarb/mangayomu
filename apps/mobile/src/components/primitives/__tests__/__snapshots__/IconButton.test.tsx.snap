// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    [
      {
        "borderRadius": 10000,
        "height": 56,
        "overflow": "hidden",
        "width": 56,
      },
      undefined,
    ]
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        {
          "alignItems": "center",
          "justifyContent": "center",
        },
      ]
    }
  >
    <Text
      name="ab-testing"
      style={
        [
          {
            "fontSize": 24,
          },
          {
            "color": "rgba(0, 0, 0, 0.87)",
          },
          {
            "textAlign": "center",
          },
        ]
      }
      type="icon"
    >
      icon
    </Text>
  </View>
</View>
`;
